### ISSUE TITLE
"Unmount skipped because path does not exist" when subPath is deleted

### ISSUE THREAD

**Issue by drigz** [kubernetes-csi/csi-driver-smb#376]
**What happened**:

I noticed that sometimes our pods get stuck in Terminating, and when I look at the kubelet logs I see:

```
mount_helper_common.go:33] Warning: Unmount skipped because path does not exist: /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb/0
nestedpendingoperations.go:301] [...] remove /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb: directory not empty"
```

**What you expected to happen**:

Pods can be deleted, even if a directory has been deleted on the smbserver.

**How to reproduce it**:

I set up a volume mount based on the examples pv-smb.yaml, pvc-smb-static.yaml, deployment.yaml, with some exceptions:

- I created `foo/bar` within the SMB share.
- I changed the deployment to run `sleep 9999` and set a subPath:

```
          volumeMounts:
            - name: smb
              mountPath: "/mnt/smb"
              readOnly: false
              subPath: foo
```

After the deployment was running, I confirmed it had mounted the SMB share, then deleted `foo/` followed by the deployment:

```
$ kubectl exec -it deploy/deployment-smb -- ls /mnt/smb                                         
bar                               
$ sudo rm -r /path/to/smb/share/foo                                                                                                                                                                  
$ kubectl delete deploy deployment-smb                                                          
deployment.apps "deployment-smb" deleted                                                                 
$ kubectl get pods                                                                              
NAME                                          READY   STATUS        RESTARTS   AGE                       
deployment-smb-67cfbbd475-q7jjb               1/1     Terminating   0          50s                       
```

The pod is stuck in terminating. kubelet shows the following errors:

```
kubelet[1558913]: W1118 21:30:46.620913 1558913 mount_helper_common.go:33] Warning: Unmount skipped because path does not exist: /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb/0
kubelet[1558913]: E1118 21:30:46.621117 1558913 nestedpendingoperations.go:301] Operation for "{volumeName:kubernetes.io/csi/smb.csi.k8s.io^unique-volumeid podName:5e50975f-6331-4e76-9968-09627864d30d nodeName:}" failed. No retries permitted until 2021-11-18 21:31:50.621060134 +0100 CET m=+801376.094888036 (durationBeforeRetry 1m4s). Error: "error cleaning subPath mounts for volume \"smb\" (UniqueName: \"kubernetes.io/csi/smb.csi.k8s.io^unique-volumeid\") pod \"5e50975f-6331-4e76-9968-09627864d30d\" (UID: \"5e50975f-6331-4e76-9968-09627864d30d\") : error deleting /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb: remove /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb: directory not empty"
```

csi-smb-node doesn't report an error - perhaps because it doesn't know about the subPath? (although there a warning "... is not a mountpoint, deleting")

```
I1118 20:28:37.399545       1 utils.go:118] GRPC call: /csi.v1.Node/NodePublishVolume
I1118 20:28:37.399595       1 utils.go:119] GRPC request: {"staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-smb/globalmount","target_path":"/var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volumes/kubernetes.io~csi/pv-smb/mount","volume_capability":{"AccessType":{"Mount":{"mount_flags":["dir_mode=0555","file_mode=0444","vers=3.0"]}},"access_mode":{"mode":5}},"volume_context":{"csi.storage.k8s.io/ephemeral":"false","csi.storage.k8s.io/pod.name":"deployment-smb-67cfbbd475-q7jjb","csi.storage.k8s.io/pod.namespace":"default","csi.storage.k8s.io/pod.uid":"5e50975f-6331-4e76-9968-09627864d30d","csi.storage.k8s.io/serviceAccount.name":"default","source":"//smb-server.default.svc.cluster.local/share"},"volume_id":"unique-volumeid"}
I1118 20:28:37.404346       1 nodeserver.go:85] NodePublishVolume: mounting /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-smb/globalmount at /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volumes/kubernetes.io~csi/pv-smb/mount with mountOptions: [bind] volumeID(unique-volumeid)
I1118 20:28:37.404403       1 mount_linux.go:175] Mounting cmd (mount) with arguments ( -o bind /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-smb/globalmount /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volumes/kubernetes.io~csi/pv-smb/mount)
I1118 20:28:37.411311       1 mount_linux.go:175] Mounting cmd (mount) with arguments ( -o bind,remount /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-smb/globalmount /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volumes/kubernetes.io~csi/pv-smb/mount)
I1118 20:28:37.414453       1 nodeserver.go:92] NodePublishVolume: mount /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-smb/globalmount at /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volumes/kubernetes.io~csi/pv-smb/mount volumeID(unique-volumeid) successfully
I1118 20:28:37.414487       1 utils.go:125] GRPC response: {}
I1118 20:28:37.512580       1 utils.go:118] GRPC call: /csi.v1.Node/NodeUnpublishVolume
I1118 20:28:37.512620       1 utils.go:119] GRPC request: {"target_path":"/var/lib/kubelet/pods/3d219b8d-4683-4ee7-8a75-525e03099cbe/volumes/kubernetes.io~csi/pv-smb/mount","volume_id":"unique-volumeid"}
I1118 20:28:37.513728       1 nodeserver.go:107] NodeUnpublishVolume: unmounting volume unique-volumeid on /var/lib/kubelet/pods/3d219b8d-4683-4ee7-8a75-525e03099cbe/volumes/kubernetes.io~csi/pv-smb/mount
I1118 20:28:37.513746       1 mount_linux.go:266] Unmounting /var/lib/kubelet/pods/3d219b8d-4683-4ee7-8a75-525e03099cbe/volumes/kubernetes.io~csi/pv-smb/mount
W1118 20:28:37.521540       1 mount_helper_common.go:133] Warning: "/var/lib/kubelet/pods/3d219b8d-4683-4ee7-8a75-525e03099cbe/volumes/kubernetes.io~csi/pv-smb/mount" is not a mountpoint, deleting
I1118 20:28:37.521655       1 nodeserver.go:112] NodeUnpublishVolume: unmount volume unique-volumeid on /var/lib/kubelet/pods/3d219b8d-4683-4ee7-8a75-525e03099cbe/volumes/kubernetes.io~csi/pv-smb/mount successfully
I1118 20:28:37.521675       1 utils.go:125] GRPC response: {}
```

If I inspect the mountpoint:

```
$ sudo ls /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb   
0
$ sudo ls -ld /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb/0
ls: cannot access '/var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb/0': No such file or directory
$ sudo umount /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb/0
(successful umount)
```

After `umount`, the pod terminates normally. I've actually installed a hacky workaround on the crontab of our clusters:

```
* * * * * journalctl -u kubelet -n 3 | sed -n "s/.*Unmount skipped because path does not exist: \(.*\)/\1/p" | xargs umount
```

**Anything else we need to know?**:

This might be a kubelet bug - I don't know much about the interface between the plugin and the kubelet itself. I'm also not sure how to reproduce without using csi-driver-smb, so not sure how it ought to be reported upstream.

I found https://github.com/kubernetes/kubernetes/issues/83089 but it just mentions the warning, not the pod stuck in Terminating. 

**Environment**:
- CSI Driver version: 1.2.0
- Kubernetes version (use `kubectl version`): 1.20.11
- OS (e.g. from /etc/os-release): Similar to Debian testing
- Kernel (e.g. `uname -a`): 5.10.46
- Install tools: helm

**Comment by andyzhangx** [comment]
related to this PR: https://github.com/kubernetes/kubernetes/pull/105348 and this issue: https://github.com/kubernetes/kubernetes/issues/97031

**Comment by andyzhangx** [comment]
this subPath issue was fixed in `1.21.9`, `1.22.6`, `1.23.2`
